src/photon/photon_algorithm.c

```diff
@@ -572,8 +572,30 @@ void dispatch_tasks(local_scheduler_state *state,
 
   /* Assign as many tasks as we can, while there are workers available. */
   DL_FOREACH_SAFE(algorithm_state->dispatch_task_queue, elt, tmp) {
-    /* TODO(atumanov): as an optimization, we can also check if all dynamic
-     * capacity is zero and bail early. */
+    /* If there is a task to assign, but there are no more available workers in
+     * the worker pool, then exit. Ensure that there will be an available
+     * worker during a future invocation of dispatch_tasks. */
+    if (utarray_len(algorithm_state->available_workers) == 0) {
+      if (utarray_len(state->child_pids) == 0) {
+        /* If there are no workers, including those pending PID registration,
+         * then we must start a new one to replenish the worker pool. */
+        start_worker(state, NIL_ACTOR_ID);
+      }
+      return; /* No workers available -- terminate early. */
+    }
+    /* Terminate early if there are no more resources available. */
```

Comment:
yeah, that's right. I assume that a task must consume some resources, even if it's a fraction of a core.